jupyterhub/scopes.py

```diff
@@ -346,7 +347,132 @@ def get_scopes_for(orm_object):
             )
         expanded_scopes = intersection
     else:
-        expanded_scopes = roles.expand_roles_to_scopes(orm_object)
+        expanded_scopes = roles.roles_to_expanded_scopes(
+            roles.get_roles_for(orm_object),
+            owner=orm_object,
+        )
+    return expanded_scopes
+
+
+def _expand_self_scope(username):
+    """
+    Users have a metascope 'self' that should be expanded to standard user privileges.
+    At the moment that is a user-filtered version (optional read) access to
+    users
+    users:name
+    users:groups
+    users:activity
+    tokens
+    servers
+    access:servers
+
+
+    Arguments:
+      username (str): user name
+
+    Returns:
+      expanded scopes (set): set of expanded scopes covering standard user privileges
+    """
+    scope_list = [
+        'read:users',
+        'read:users:name',
+        'read:users:groups',
+        'users:activity',
+        'read:users:activity',
+        'servers',
+        'delete:servers',
+        'read:servers',
+        'tokens',
+        'read:tokens',
+        'access:servers',
+    ]
+    return {f"{scope}!user={username}" for scope in scope_list}
+
+
+def _expand_scope(scope):
+    """Returns a scope and all all subscopes
+
+    Arguments:
+      scope (str): the scope to expand
+
+    Returns:
+      expanded scope (set): set of all scope's subscopes including the scope itself
+    """
+
+    # remove filter, save for later
+    scope_name, sep, filter_ = scope.partition('!')
+
+    # expand scope and subscopes
+    expanded_scope_names = set()
+
+    def _add_subscopes(scope_name):
+        expanded_scope_names.add(scope_name)
+        if scope_definitions[scope_name].get('subscopes'):
+            for subscope in scope_definitions[scope_name].get('subscopes'):
+                _add_subscopes(subscope)
+
+    _add_subscopes(scope_name)
+
+    # reapply !filter
+    if filter_:
+        expanded_scopes = {
+            f"{scope_name}!{filter_}" for scope_name in expanded_scope_names
+        }
+        # special handling of server filter
+        # any access via server filter includes permission to read the user's name
+        resource, _, value = filter_.partition('=')
+        if resource == 'server':
+            username, _, server = value.partition('/')
+            expanded_scopes.add(f'read:users:name!user={username}')
+    else:
+        expanded_scopes = expanded_scope_names
+
+    return expanded_scopes
+
+
+def expand_scopes(scopes, owner=None):
+    """Returns a set of fully expanded scopes for a collection of raw scopes
```

Comment:
raw scopes, metascopes, and expanded scopes.

This is a bit confusing to me, and I think I could benefit from a clarification of the terminology.

Raw scopes is not yet expanded scopes I guess, which could result in the same scopes or not, depending on if they are metascopes?